modules/core/src/lifecycle/component-state.js

```diff
@@ -0,0 +1,179 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// TODO - hack for initial PR, remove
+const ASYNC_PROPS = {
+  data: null
+};
+
+export default class ComponentState {
+  constructor(object = null) {
+    this.object = object;
+    this.asyncProps = {}; // Prop values that the layer sees
+    this.onAsyncPropUpdated = () => {};
+  }
+
+  //
+  // ASYNC PROP HANDLING
+  //
+
+  // Checks if a prop is overridden
+  hasAsyncProp(propName) {
+    return propName in this.asyncProps;
+  }
+
+  // Returns value of an overriden prop
+  getAsyncProp(propName) {
+    const asyncProp = this.asyncProps[propName];
+    return asyncProp && asyncProp.resolvedValue;
+  }
+
+  isAsyncPropLoading(propName) {
+    const asyncProp = this.asyncProps[propName];
+    return (
+      asyncProp &&
+      asyncProp.pendingLoadCount > 0 &&
+      asyncProp.pendingLoadCount !== asyncProp.resolvedLoadCount
+    );
+  }
+
+  // Updates all async/overridden props (when new props come in)
+  // Checks if urls have changed, starts loading, or removes override
+  setAsyncProps(props) {
+    // TODO - use async props from the layer's prop types
+    for (const propName in ASYNC_PROPS) {
+      // const value = this.layer.props._getOriginalValue(propName)
+      if (propName in props) {
+        const value = props[propName];
+        this._updateAsyncProp(propName, value);
+      }
+    }
+  }
+
+  // Intercept strings (URLs) and Promises and activates loading and prop rewriting
+  _updateAsyncProp(propName, value) {
+    // Makes sure a record exists for this prop
+    this._getAsyncPropData(propName);
+
+    if (value instanceof Promise) {
+      this._loadPromiseProp(propName, value);
+      return;
+    }
+
+    if (typeof value === 'string') {
+      this._loadUrlProp(propName, value);
+      return;
+    }
+
+    // else, normal, non-async value. Just store value for now
+    this._setPropValue(propName, value);
+  }
+
+  // interpret value string as promise and track the "loading"
+  _loadPromiseProp(propName, promise) {
+    if (!this._didAsyncInputValueChange(propName, promise)) {
```

Comment:
Move into `_updateAsyncProp`?

Reply:
> Move into _updateAsyncProp?

Done. I went back and forth, but the way the code turned out, it makes sense to join these functions.